packages/@lwc/engine-core/src/3rdparty/snabbdom/snabbdom.ts

```diff
@@ -139,7 +152,6 @@ export function updateDynamicChildren(parentElm: Node, oldCh: VNodes, newCh: VNo
                         newStartVnode.hook.insert(newStartVnode, parentElm, oldStartVnode.elm!);
                     } else {
                         patchVnode(elmToMove, newStartVnode);
-                        oldCh[idxInOld] = undefined as any;
```

Comment:
Due to `VNodes` being read-only now, I had to remove this line. AFAICT, it is not needed. I don't understand why we would need to set this to `undefined` in the `oldCh` array. The old children should just be garbage-collected, as they either belong to the old VNode:

https://github.com/salesforce/lwc/blob/c26277bf44170ab5ba02cd7775e1cff3e7df86e0/packages/%40lwc/engine-core/src/framework/hooks.ts#L179

Or are replaced on the current VM:

https://github.com/salesforce/lwc/blob/c26277bf44170ab5ba02cd7775e1cff3e7df86e0/packages/%40lwc/engine-core/src/framework/vm.ts#L418-L419

https://github.com/salesforce/lwc/blob/c26277bf44170ab5ba02cd7775e1cff3e7df86e0/packages/%40lwc/engine-core/src/framework/vm.ts#L444

Reply:
I don't yet fully understand the implications of removing this line here. The fact that it breaks no test is good news to me.

Reply:
You CANNOT remove this line! :)

Basically, this line actuates when there is a new element who's key is found on the old vnodes, there are 2 possibilities:

1. the old node is in a position prior to the new position.
2. the old node is in a position after the new position.

The posibility of being in the same position is not possible in this branch of the code.

On top of that, there is another problem: "whether or not the tagName (sel) matches.

So, the problem is that if the sel matches, and the position is prior the new position (remember that this algo goes backward), since we are in a loop, there is a possibility that later in the game you encounter another element that must be inserted in the position marked by the old position, and the diffing algo tries to remove the element associated to the old vnodes.

My hunch is that this will result on the moved node to disappear from the DOM, and it is very hard to write a test that validate this assumption, but we can certainly try.

Reply:
@caridy Thanks for the explanation. I'm looking into it, and it turns out that this `else` block is never hit in either our Jest tests or Karma tests:

https://github.com/salesforce/lwc/blob/d93e2963d8d32b138ce9ed130d98e7bef84b3d74/packages/%40lwc/engine-core/src/3rdparty/snabbdom/snabbdom.ts#L127-L146

In other words, `isUndef(idxInOld)` is always true. I'll try to write a test where it is false.

Reply:
OK, I managed to find a Karma test to hit this line of code (I'll open a separate PR to improve our test coverage). Based on the usage patterns, I think we can do this:

1. Instead of using `oldCh` directly, use a clone (`[...oldCh]`) inside of `updateDynamicChildren`. AFAICT, `oldCh` is only used inside of that function – afterwards, it is discard and GC'ed.
2. To avoid the perf hit of cloning arrays, we can do the clone only on line 142 (i.e. when calling `oldCh[idxInOld] = undefined as any; `).

@caridy How does this sound?

Reply:
I'm not sure the clone solves the problem, unless you always rely on the cloned array for the different paths of the algo.

Reply:
also closing via `[...oldCh]` will trigger the iterable protocol that is not ideal.

Reply:
@caridy I'm saying we should do:

```diff
- oldCh[idxInOld] = undefined as any;
+ const oldChClone = [...oldCh];
+ oldChClone[idxInOld] = undefined as any;
+ oldCh = oldChClone;
```

It solves the problem because the `oldCh` array is `Readonly` – although only because of the empty array `[]`, which may be shared between VNodes due to the optimization in this PR. (A non-empty array cannot be hoisted because it will contain VNodes, not static objects. This is why we only need a shallow clone, not a deep clone.)

> also closing via `[...oldCh]` will trigger the iterable protocol that is not ideal.

Can you elaborate? Are you concerned about performance? It seems to me that this code path is very infrequent, so I'm not really concerned about the perf implications there.

Reply:
Ok, I'm fine with this approach. In locker code, we try to stay away from `[...array]` due to the perf implications of the iterables. cc @jdalton, but as you said, this is a very rare code path anyways.